crates/vm/src/expr.rs

```diff
@@ -273,57 +274,53 @@ impl From<Query> for Option<ColumnOp> {
     }
 }
 
-#[derive(Debug, PartialEq, Eq, Clone, Default)]
-#[repr(transparent)]
-/// A set of [`MemTable`]s referenced by a query plan by their [`SourceId`].
+/// A set of [`SourceTable`]s referenced by a query plan by their [`SourceId`].
 ///
-/// Rather than embedding [`MemTable`]s in query plans, we store a `SourceExpr::MemTable`,
+/// Rather than embedding tables in query plans, we store a `SourceExpr::MemTable`,
 /// which contains the information necessary for optimization along with a [`SourceId`].
 /// Query execution then executes the plan, and when it encounters a `SourceExpr::MemTable`,
-/// retrieves the [`MemTable`] from the corresponding `SourceSet`.
+/// retrieves the `T` table from the corresponding `SourceSet`.
 /// This allows query plans to be re-used, though each execution requires a new `SourceSet`.
 ///
-/// Internally, the `SourceSet` stores an `Option<MemTable>` for each planned [`SourceId`].
+/// Internally, the `SourceSet` stores an `Option<T>` for each planned [`SourceId`].
 /// During execution, the VM will [`Option::take`] the [`MemTable`] to consume them.
 /// This means that a query plan may not include multiple references to the same [`SourceId`].
-pub struct SourceSet(Vec<Option<MemTable>>);
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[repr(transparent)]
+pub struct SourceSet<T, const N: usize>(ArrayVec<Option<T>, N>);
```

Comment:
Why the new `const N: usize` parameter? Do we have evidence that `ArrayVec` is meaningfully faster here than just `Vec`? Propagating the const parameter through everything that uses `SourceSet` adds a lot of noise to our code which I'd prefer to avoid without a good reason.